@thockin  I'm on OSX, so, probably ignore me. It worked once, but at the moment I consistently see 
```
.make/_output/bin/deepcopy-gen.mk:2: *** missing separator.  Stop.
make: *** [generated_files] Error 2
```
for all invocations of make.

--- a/Makefile
+++ b/Makefile
@@ -14,7 +14,7 @@
 
 DBG_MAKEFILE ?=
 ifeq ($(DBG_MAKEFILE),1)
-    $(warning ***** starting makefile for goal(s) "$(MAKECMDGOALS)")
+    $(warning ***** starting Makefile for goal(s) "$(MAKECMDGOALS)")
     $(warning ***** $(shell date))
 else
     # If we're not debugging the Makefile, don't echo recipes.
@@ -29,7 +29,7 @@ endif
 #   test: Run tests.
 #   clean: Clean up.
 
-# It's necessary to set this because some docker images don't make sh -> bash.
+# It's necessary to set this because some environments don't link sh -> bash.
 SHELL := /bin/bash
 
 # We don't need make's built-in rules.
@@ -37,16 +37,17 @@ MAKEFLAGS += --no-builtin-rules
 .SUFFIXES:
 
 # Constants used throughout.
+.EXPORT_ALL_VARIABLES:
 OUT_DIR ?= _output
 BIN_DIR := $(OUT_DIR)/bin
 PRJ_SRC_PATH := k8s.io/kubernetes
+GENERATED_FILE_PREFIX := zz_generated.
 
 # Metadata for driving the build lives here.
 META_DIR := .make
 
-export KUBE_GOFLAGS := $(GOFLAGS)
-
-export KUBE_GOLDFLAGS := $(GOLDFLAGS)
+KUBE_GOFLAGS := $(GOFLAGS)
+KUBE_GOLDFLAGS := $(GOLDFLAGS)
 
 # Build code.
 #
@@ -184,6 +185,14 @@ clean: clean_meta
 clean_meta:
 	rm -rf $(META_DIR)
 
+# Remove all auto-generated artifacts.
+#
+# Example:
+#   make clean_generated
+.PHONY: clean_generated
+clean_generated:
+	find . -type f -name $(GENERATED_FILE_PREFIX)\* | xargs rm -f
+
 # Run 'go vet'.
 #
 # Args:
@@ -230,5 +239,10 @@ cross:
 $(notdir $(abspath $(wildcard cmd/*/))): generated_files
 	hack/make-rules/build.sh cmd/$@
 
-# Include logic for generated files.
-include Makefile.generated_files
+# Produce auto-generated files needed for the build.
+#
+# Example:
+#   make generated_files
+.PHONY: generated_files
+generated_files:
+	$(MAKE) -f Makefile.$@ $@

--- a/Makefile.generated_files
+++ b/Makefile.generated_files
@@ -108,24 +108,11 @@ generated_files: gen_deepcopy gen_conversion
 ifeq ($(DBG_MAKEFILE),1)
     $(warning ***** finding all *.go dirs)
 endif
-ALL_GO_DIRS := $(shell             \
-    find .                         \
-        -not \(                    \
-            \(                     \
-                -path ./vendor -o  \
-                -path ./_\* -o     \
-                -path ./.\* -o     \
-                -path ./docs -o    \
-                -path ./examples   \
-            \) -prune              \
-        \)                         \
-        -type f -name \*.go        \
-        | sed 's|^./||'            \
-        | xargs -n1 dirname        \
-        | sort -u                  \
+ALL_GO_DIRS := $(shell                                                   \
+    hack/make-rules/helpers/cache_go_dirs.sh $(META_DIR)/all_go_dirs.mk  \
 )
 
-# The name of the make metadata file listing Go files.
+# The name of the metadata file which lists *.go files in each pkg.
 GOFILES_META := gofiles.mk
 
 # Establish a dependency between the deps file and the dir.  Whenever a dir
@@ -314,7 +301,7 @@ CONVERSION_FILENAME := $(CONVERSION_BASENAME).go
 # The tool used to generate conversions.
 CONVERSION_GEN := $(BIN_DIR)/conversion-gen
 
-# The name of the make metadata file controlling conversions.
+# The name of the metadata file listing conversion peers for each pkg.
 CONVERSIONS_META := conversions.mk
 
 # All directories that request any form of conversion generation.

--- a/Makefile.generated_files
+++ b/Makefile.generated_files
@@ -12,8 +12,24 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-# Constants used throughout.
-GENERATED_FILE_PREFIX := zz_generated.
+# Don't allow an implicit 'all' rule.  This is not a user-facing file.
+ifeq ($(MAKECMDGOALS),)
+    $(error This Makefile requires an explicit rule to be specified)
+endif
+
+ifeq ($(DBG_MAKEFILE),1)
+    $(warning ***** starting Makefile.generated_files for goal(s) "$(MAKECMDGOALS)")
+    $(warning ***** $(shell date))
+endif
+
+
+# It's necessary to set this because some environments don't link sh -> bash.
+SHELL := /bin/bash
+
+# This rule collects all the generated file sets into a single rule.  Other
+# rules should depend on this to ensure generated files are rebuilt.
+.PHONY: generated_files
+generated_files: gen_deepcopy gen_conversion
 
 # Code-generation logic.
 #
@@ -455,17 +471,3 @@ sinclude $(META_DIR)/$(CONVERSION_GEN).mk
 $(CONVERSION_GEN):
 	hack/make-rules/build.sh cmd/libs/go2idl/conversion-gen
 	touch $@
-
-# This rule collects all the generated file sets into a single dep, which is
-# defined BELOW the *_FILES variables and leaves higher-level rules clean.
-# Top-level rules should depend on this to ensure generated files are rebuilt.
-.PHONY: generated_files
-generated_files: gen_deepcopy gen_conversion
-
-# Remove all auto-generated artifacts.
-#
-# Example:
-#   make clean_generated
-.PHONY: clean_generated
-clean_generated:
-	find . -type f -name $(GENERATED_FILE_PREFIX)\* | xargs rm -f

--- a/hack/make-rules/helpers/cache_go_dirs.sh
+++ b/hack/make-rules/helpers/cache_go_dirs.sh
@@ -1,0 +1,65 @@
+#!/bin/bash
+
+# Copyright 2014 The Kubernetes Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# This script finds, caches, and prints a list of all directories that hold
+# *.go files.  If any directory is newer than the cache, re-find everything and
+# update the cache.  Otherwise use the cached file.
+
+set -o errexit
+set -o nounset
+set -o pipefail
+
+if [[ -z "${1:-}" ]]; then
+    echo "usage: $0 <cache-file>"
+    exit 1
+fi
+CACHE="$1"; shift
+
+# This is a partial 'find' command.  The caller is expected to pass the
+# remaining arguments.
+#
+# Example:
+#   kfind -type f -name foobar.go
+function kfind() {
+    find .                         \
+        -not \(                    \
+            \(                     \
+                -path ./vendor -o  \
+                -path ./_\* -o     \
+                -path ./.\* -o     \
+                -path ./docs -o    \
+                -path ./examples   \
+            \) -prune              \
+        \)                         \
+        "$@"
+}
+
+NEED_FIND=true
+# It's *significantly* faster to check whether any directories are newer than
+# the cache than to blindly rebuild it.
+if [[ -f "${CACHE}" ]]; then
+    N=$(kfind -type d -newer "${CACHE}" -print -quit | wc -l)
+    [[ "${N}" == 0 ]] && NEED_FIND=false
+fi
+mkdir -p $(dirname "${CACHE}")
+if $("${NEED_FIND}"); then
+    kfind -type f -name \*.go  \
+        | xargs -n1 dirname    \
+        | sort -u              \
+        | sed 's|^./||'        \
+        > "${CACHE}"
+fi
+cat "${CACHE}"

